lib/resty/limit/count.lua

```diff
@@ -0,0 +1,83 @@
+-- implement GitHub request rate limiting: https://developer.github.com/v3/#rate-limiting
+
+local ngx_shared = ngx.shared
+local setmetatable = setmetatable
+local assert = assert
+
+
+local _M = {
+   _VERSION = '0.01'
+}
+
+
+local mt = {
+    __index = _M
+}
+
+
+-- the "limit" argument controls number of request allowed in a time window.
+-- time "window" argument controls the time window in seconds.
+function _M.new(dict_name, limit, window)
+    local dict = ngx_shared[dict_name]
+    if not dict then
+        return nil, "shared dict not found"
+    end
+
+    assert(limit > 0 and window > 0)
+
+    local self = {
+        dict = dict,
+        limit = limit,
+        window = window,
+    }
+
+    return setmetatable(self, mt)
+end
+
+
+function _M.incoming(self, key, commit)
+    local dict = self.dict
+    local limit = self.limit
+    local window = self.window
+
+    local remaining, err
+
+    if commit then
+        remaining, err = dict:incr(key, -1, limit)
+        if not remaining then
+            return nil, err
+        end
+        if remaining == limit - 1 then
+            local ok, err = dict:expire(key, window)
+            if not ok then
+                return nil, err
+            end
+        end
+
+    else
+        remaining = (dict:get(key) or 0) - 1
+    end
+
+    if remaining < 0 then
+        return nil, "rejected"
+    end
+
+    return 0, remaining
+end
+
+
+-- uncommit remaining and return remaining value
+function _M.uncommit(self, key)
+    assert(key)
+    local dict = self.dict
+
+    local remaining, err = dict:incr(key, 1)
```

Comment:
There is a small race condition here between `commit()` and `uncommit` in which time window the key is expired. Maybe we should not return an error here when the err is "not found"?

Reply:
Yes I think we should just `get` again and return whatever the new value is. The old window has expired and no need to `incr` it anymore.

Reply:
@agentzh @dndx fixed race condition. please have a look.